tools/lsp/common/rename_component.rs

```diff
@@ -329,310 +268,491 @@ fn fix_exports(
-        validity_range: &TextRange,
-        old_type: &SmolStr,
-        new_type: &str,
-        edits: &mut Vec<common::SingleTextEdit>,
-    ) {
-        visit_document_components(document_node, &mut move |component| {
-            if validity_range.contains_range(component.text_range()) {
-                replace_element_types(
-                    document_cache,
-                    &component.Element(),
-                    old_type,
-                    new_type,
-                    edits,
-                );
+#[derive(Clone, Debug)]
+struct DeclarationNodeQuery {
+    info: common::token_info::TokenInfo,
+    name: SmolStr,
+    token: SyntaxToken,
+}
+
+impl DeclarationNodeQuery {
+    fn new(document_cache: &common::DocumentCache, token: SyntaxToken) -> Option<Self> {
+        let info = common::token_info::token_info(document_cache, token.clone())?;
+        let name = i_slint_compiler::parser::normalize_identifier(token.text());
+
+        Some(DeclarationNodeQuery { info, name, token })
+    }
+
+    fn sub_query(&self, token: SyntaxToken) -> Option<Self> {
+        let name = i_slint_compiler::parser::normalize_identifier(token.text());
+
+        Some(DeclarationNodeQuery { info: self.info.clone(), name, token })
+    }
+
+    fn is_export_identifier_or_external_name(&self) -> bool {
+        self.token.kind() == SyntaxKind::Identifier
+            && [SyntaxKind::ExportIdentifier, SyntaxKind::ExternalName]
+                .contains(&self.token.parent().kind())
+    }
+
+    fn start_token(&self) -> Option<SyntaxToken> {
+        if self.is_export_identifier_or_external_name() {
+            None
+        } else {
+            Some(self.token.clone())
+        }
+    }
+
+    fn is_same_symbol(&self, document_cache: &common::DocumentCache, token: SyntaxToken) -> bool {
+        let Some(info) = common::token_info::token_info(document_cache, token) else {
+            return false;
+        };
+
+        match (&self.info, &info) {
+            (common::token_info::TokenInfo::Type(s), common::token_info::TokenInfo::Type(o)) => {
+                s == o
             }
-        });
+            (
+                common::token_info::TokenInfo::ElementType(s),
+                common::token_info::TokenInfo::ElementType(o),
+            ) => s == o,
+            (
+                common::token_info::TokenInfo::ElementRc(s),
+                common::token_info::TokenInfo::ElementRc(o),
+            ) => Rc::ptr_eq(s, o),
+            (
+                common::token_info::TokenInfo::NamedReference(s),
+                common::token_info::TokenInfo::NamedReference(o),
+            ) => s == o,
+            (
+                common::token_info::TokenInfo::EnumerationValue(s),
+                common::token_info::TokenInfo::EnumerationValue(o),
+            ) => s == o,
+            (
+                common::token_info::TokenInfo::LocalProperty(s),
+                common::token_info::TokenInfo::LocalProperty(o),
```

Comment:
I wonder if LocalProperty and LocalCallback can also be the same as a NamedReference or IncompleteNamedReference

Reply:
I tired to come up with some valid comparison function and probably did some useless ones...

E.g. for the property renaming I played with yesterday evening I need to compare a `LocalProperty` with `NamedReference`. I'll just remove all the comparisons not yet used (and tested ;-).